Use of ‘auto’ in parameter declaration doesn't comply with the C++17 standard

### Bug description
Some functions use `auto` in parameter declaration, which is only available with ‘-std=c++20’ or ‘-fconcepts’.
But ALFI is currently intended to comply the C++17 standard.
The problematic functions must be refactored. 

### Steps to Reproduce
1. Change the `CMAKE_CXX_STANDARD` in `examples/dist_gnuplot/CMakeLists.txt` from `23` to `17`.
2. Compile the target `dist_gnuplot`.
3. See the output.

The output (project directory is replaced with `<project-dir>`:
```text
====================[ Build | dist_gnuplot | FastParallel ]=====================
cmake --build <project-dir>/build/fastparallel --target dist_gnuplot -j 14
[0/2] Re-checking globbed directories...
[1/3] Building CXX object examples/dist_gnuplot/CMakeFiles/dist_gnuplot.dir/dist_gnuplot.cpp.o
In file included from <project-dir>/ALFI/./ALFI/dist.h:6,
                 from <project-dir>/ALFI/./ALFI.h:5,
                 from <project-dir>/examples/dist_gnuplot/dist_gnuplot.cpp:3:
<project-dir>/ALFI/./ALFI/util/points.h:7:22: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
    7 |         void lin_map(auto& points, Number a, Number b, Number c, Number d) {
      |                      ^~~~
<project-dir>/ALFI/./ALFI/util/points.h:17:31: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   17 |         auto lin_mapped(const auto& points, Number a, Number b, Number c, Number d) {
      |                               ^~~~
<project-dir>/ALFI/./ALFI/util/points.h:24:22: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   24 |         void stretch(auto& points, Number a, Number b) {
      |                      ^~~~
<project-dir>/ALFI/./ALFI/util/points.h:38:30: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   38 |         auto stretched(const auto& points, Number a, Number b) {
      |                              ^~~~
In file included from <project-dir>/ALFI/./ALFI.h:6:
<project-dir>/ALFI/./ALFI/misc.h:13:31: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   13 |                         const auto& X,
      |                               ^~~~
<project-dir>/ALFI/./ALFI/misc.h:14:31: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   14 |                         const auto& Y,
      |                               ^~~~
<project-dir>/ALFI/./ALFI/misc.h:15:31: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   15 |                         const auto& xx,
      |                               ^~~~
In file included from <project-dir>/ALFI/./ALFI.h:7:
<project-dir>/ALFI/./ALFI/poly.h:12:19: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   12 |         val(const auto& coeffs, Number x) {
      |                   ^~~~
<project-dir>/ALFI/./ALFI/poly.h:22:19: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   22 |         val(const auto& coeffs, const ContainerX& X) {
      |                   ^~~~
<project-dir>/ALFI/./ALFI/poly.h:33:42: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   33 |         Container<Number> lagrange(const auto& X, const auto& Y) {
      |                                          ^~~~
<project-dir>/ALFI/./ALFI/poly.h:33:57: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   33 |         Container<Number> lagrange(const auto& X, const auto& Y) {
      |                                                         ^~~~
<project-dir>/ALFI/./ALFI/poly.h:78:47: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   78 |         Container<Number> lagrange_vals(const auto& X, const auto& Y, const auto& xx) {
      |                                               ^~~~
<project-dir>/ALFI/./ALFI/poly.h:78:62: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   78 |         Container<Number> lagrange_vals(const auto& X, const auto& Y, const auto& xx) {
      |                                                              ^~~~
<project-dir>/ALFI/./ALFI/poly.h:78:77: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   78 |         Container<Number> lagrange_vals(const auto& X, const auto& Y, const auto& xx) {
      |                                                                             ^~~~
<project-dir>/ALFI/./ALFI/poly.h:123:46: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  123 |         Container<Number> imp_lagrange(const auto& X, const auto& Y) {
      |                                              ^~~~
<project-dir>/ALFI/./ALFI/poly.h:123:61: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  123 |         Container<Number> imp_lagrange(const auto& X, const auto& Y) {
      |                                                             ^~~~
<project-dir>/ALFI/./ALFI/poly.h:183:51: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  183 |         Container<Number> imp_lagrange_vals(const auto& X, const auto& Y, const auto& xx, Number epsilon = std::numeric_limits<Number>::epsilon()) {
      |                                                   ^~~~
<project-dir>/ALFI/./ALFI/poly.h:183:66: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  183 |         Container<Number> imp_lagrange_vals(const auto& X, const auto& Y, const auto& xx, Number epsilon = std::numeric_limits<Number>::epsilon()) {
      |                                                                  ^~~~
<project-dir>/ALFI/./ALFI/poly.h:183:81: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  183 |         Container<Number> imp_lagrange_vals(const auto& X, const auto& Y, const auto& xx, Number epsilon = std::numeric_limits<Number>::epsilon()) {
      |                                                                                 ^~~~
<project-dir>/ALFI/./ALFI/poly.h:241:40: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  241 |         Container<Number> newton(const auto& X, const auto& Y) {
      |                                        ^~~~
<project-dir>/ALFI/./ALFI/poly.h:241:55: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  241 |         Container<Number> newton(const auto& X, const auto& Y) {
      |                                                       ^~~~
<project-dir>/ALFI/./ALFI/poly.h:294:45: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  294 |         Container<Number> newton_vals(const auto& X, const auto& Y, const auto& xx) {
      |                                             ^~~~
<project-dir>/ALFI/./ALFI/poly.h:294:60: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  294 |         Container<Number> newton_vals(const auto& X, const auto& Y, const auto& xx) {
      |                                                            ^~~~
<project-dir>/ALFI/./ALFI/poly.h:294:75: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  294 |         Container<Number> newton_vals(const auto& X, const auto& Y, const auto& xx) {
      |                                                                           ^~~~
In file included from <project-dir>/ALFI/./ALFI/spline.h:4,
                 from <project-dir>/ALFI/./ALFI.h:8:
<project-dir>/ALFI/./ALFI/spline/linear.h:14:63: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   14 |                 static Container<Number> compute_coeffs(const auto& X, const auto& Y) {
      |                                                               ^~~~
<project-dir>/ALFI/./ALFI/spline/linear.h:14:78: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   14 |                 static Container<Number> compute_coeffs(const auto& X, const auto& Y) {
      |                                                                              ^~~~
<project-dir>/ALFI/./ALFI/spline/linear.h:50:56: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   50 |                 LinearSpline(ContainerXType&& X, const auto& Y) {
      |                                                        ^~~~
<project-dir>/ALFI/./ALFI/spline/linear.h:61:58: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   61 |                 void construct(ContainerXType&& X, const auto& Y) {
      |                                                          ^~~~
In file included from <project-dir>/ALFI/./ALFI/spline/quadratic.h:9,
                 from <project-dir>/ALFI/./ALFI/spline.h:5:
<project-dir>/ALFI/./ALFI/spline/../util/arrays.h:43:38: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   43 |         Container<Number> diff(const auto& container) {
      |                                      ^~~~
In file included from <project-dir>/ALFI/./ALFI/spline/../util/spline.h:6,
                 from <project-dir>/ALFI/./ALFI/spline/quadratic.h:11:
<project-dir>/ALFI/./ALFI/spline/../util/../util/linalg.h:101:56: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  101 |         Container<Number> tridiag_solve_unstable(const auto& lower, auto&& diag, const auto& upper, auto&& right) {
      |                                                        ^~~~
<project-dir>/ALFI/./ALFI/spline/../util/../util/linalg.h:101:69: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  101 |         Container<Number> tridiag_solve_unstable(const auto& lower, auto&& diag, const auto& upper, auto&& right) {
      |                                                                     ^~~~
<project-dir>/ALFI/./ALFI/spline/../util/../util/linalg.h:101:88: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  101 |         Container<Number> tridiag_solve_unstable(const auto& lower, auto&& diag, const auto& upper, auto&& right) {
      |                                                                                        ^~~~
<project-dir>/ALFI/./ALFI/spline/../util/../util/linalg.h:101:101: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  101 |         Container<Number> tridiag_solve_unstable(const auto& lower, auto&& diag, const auto& upper, auto&& right) {
      |                                                                                                     ^~~~
<project-dir>/ALFI/./ALFI/spline/../util/../util/linalg.h:143:41: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  143 |         Container<Number> tridiag_solve(auto&& lower, auto&& diag, auto&& upper, auto&& right) {
      |                                         ^~~~
<project-dir>/ALFI/./ALFI/spline/../util/../util/linalg.h:143:55: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  143 |         Container<Number> tridiag_solve(auto&& lower, auto&& diag, auto&& upper, auto&& right) {
      |                                                       ^~~~
<project-dir>/ALFI/./ALFI/spline/../util/../util/linalg.h:143:68: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  143 |         Container<Number> tridiag_solve(auto&& lower, auto&& diag, auto&& upper, auto&& right) {
      |                                                                    ^~~~
<project-dir>/ALFI/./ALFI/spline/../util/../util/linalg.h:143:82: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  143 |         Container<Number> tridiag_solve(auto&& lower, auto&& diag, auto&& upper, auto&& right) {
      |                                                                                  ^~~~
<project-dir>/ALFI/./ALFI/spline/quadratic.h:137:63: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  137 |                 static Container<Number> compute_coeffs(const auto& X, const auto& Y, Type type = typename Types::Default{}) {
      |                                                               ^~~~
<project-dir>/ALFI/./ALFI/spline/quadratic.h:137:78: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  137 |                 static Container<Number> compute_coeffs(const auto& X, const auto& Y, Type type = typename Types::Default{}) {
      |                                                                              ^~~~
<project-dir>/ALFI/./ALFI/spline/quadratic.h:268:59: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  268 |                 QuadraticSpline(ContainerXType&& X, const auto& Y, Type type = typename Types::Default{}) {
      |                                                           ^~~~
<project-dir>/ALFI/./ALFI/spline/quadratic.h:279:58: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  279 |                 void construct(ContainerXType&& X, const auto& Y, Type type = typename Types::Default{}) {
      |                                                          ^~~~
In file included from <project-dir>/ALFI/./ALFI/spline.h:6:
<project-dir>/ALFI/./ALFI/spline/cubic.h:90:63: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   90 |                 static Container<Number> compute_coeffs(const auto& X, const auto& Y, Type type = typename Types::Default{}) {
      |                                                               ^~~~
<project-dir>/ALFI/./ALFI/spline/cubic.h:90:78: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
   90 |                 static Container<Number> compute_coeffs(const auto& X, const auto& Y, Type type = typename Types::Default{}) {
      |                                                                              ^~~~
<project-dir>/ALFI/./ALFI/spline/cubic.h:399:55: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  399 |                 CubicSpline(ContainerXType&& X, const auto& Y, Type type = typename Types::Default{}) {
      |                                                       ^~~~
<project-dir>/ALFI/./ALFI/spline/cubic.h:410:58: warning: use of ‘auto’ in parameter declaration only available with ‘-std=c++20’ or ‘-fconcepts’
  410 |                 void construct(ContainerXType&& X, const auto& Y, Type type = typename Types::Default{}) {
      |                                                          ^~~~
[2/3] Linking CXX executable examples/dist_gnuplot/dist_gnuplot

Build finished
```

### Expected behavior
Expected to compile without warnings.

### Problematic functions
From the compiler output:
- `lin_map`
- `lin_mapped`
- `stretch`
- `stretched`
- `barycentric`
- `val` (for single value)
- `val` (for container)
- `lagrange`
- `lagrange_vals`
- `imp_lagrange`
- `imp_lagrange_vals`
- `newton`
- `newton_vals`
- `LinearSpline::compute_coeffs`
- `LinearSpline::LinearSpline`
- `LinearSpline::construct`
- `diff`
- `tridiag_solve_unstable`
- `tridiag_solve`
- `QuadraticSpline::compute_coeffs`
- `QuadraticSpline::QuadraticSpline`
- `QuadraticSpline::construct`
- `CubicSpline::compute_coeffs`
- `CubicSpline::CubicSpline`
- `CubicSpline::construct`

### Release
- Version: 0.0.0.0
- OS: Kubuntu 24.10

--- a/ALFI/ALFI/dist.h
+++ b/ALFI/ALFI/dist.h
@@ -107,7 +107,7 @@ namespace alfi::dist {
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
 	Container<Number> chebyshev_stretched(SizeT n, Number a, Number b) {
-		return points::stretched(chebyshev(n, a, b), a, b);
+		return points::stretched<Number,Container>(chebyshev(n, a, b), a, b);
 	}
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
@@ -126,7 +126,7 @@ namespace alfi::dist {
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
 	Container<Number> chebyshev_ellipse_stretched(SizeT n, Number a, Number b, Number ratio) {
-		return points::stretched(chebyshev_ellipse(n, a, b, ratio), a, b);
+		return points::stretched<Number,Container>(chebyshev_ellipse(n, a, b, ratio), a, b);
 	}
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
@@ -250,7 +250,7 @@ namespace alfi::dist {
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
 	Container<Number> erf_stretched(SizeT n, Number a, Number b, Number steepness) {
-		return points::stretched(erf(n, a, b, steepness), a, b);
+		return points::stretched<Number,Container>(erf(n, a, b, steepness), a, b);
 	}
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>

--- a/ALFI/ALFI/misc.h
+++ b/ALFI/ALFI/misc.h
@@ -10,9 +10,9 @@ namespace alfi::misc {
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
 	Container<Number> barycentric(
-			const auto& X,
-			const auto& Y,
-			const auto& xx,
+			const Container<Number>& X,
+			const Container<Number>& Y,
+			const Container<Number>& xx,
 			dist::Type dist_type = dist::Type::GENERAL,
 			Number epsilon = std::numeric_limits<Number>::epsilon()
 	) {

--- a/ALFI/ALFI/poly.h
+++ b/ALFI/ALFI/poly.h
@@ -7,9 +7,9 @@
 #include "util/numeric.h"
 
 namespace alfi::poly {
-	template <typename Number = DefaultNumber>
+	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
 	std::enable_if_t<!traits::has_size<Number>::value, Number>
-	val(const auto& coeffs, Number x) {
+	val(const Container<Number>& coeffs, Number x) {
 		Number result = 0;
 		for (const Number& c : coeffs) {
 			result = x * result + c;
@@ -17,9 +17,9 @@ namespace alfi::poly {
 		return result;
 	}
 
-	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer, typename ContainerX = Container<Number>>
-	std::enable_if_t<traits::has_size<ContainerX>::value, Container<Number>>
-	val(const auto& coeffs, const ContainerX& X) {
+	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
+	std::enable_if_t<traits::has_size<Container<Number>>::value, Container<Number>>
+	val(const Container<Number>& coeffs, const Container<Number>& X) {
 		Container<Number> result(X.size(), 0);
 #if defined(_OPENMP) && !defined(ALFI_DISABLE_OPENMP)
 #pragma omp parallel for
@@ -33,7 +33,7 @@ namespace alfi::poly {
 	}
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
-	Container<Number> lagrange(const auto& X, const auto& Y) {
+	Container<Number> lagrange(const Container<Number>& X, const Container<Number>& Y) {
 		if (X.size() != Y.size()) {
 			std::cerr << "Error in function " << __FUNCTION__
 					  << ": Vectors X (of size " << X.size()
@@ -78,7 +78,7 @@ namespace alfi::poly {
 	}
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
-	Container<Number> lagrange_vals(const auto& X, const auto& Y, const auto& xx) {
+	Container<Number> lagrange_vals(const Container<Number>& X, const Container<Number>& Y, const Container<Number>& xx) {
 		const auto nn = xx.size();
 
 		if (X.size() != Y.size()) {
@@ -123,7 +123,7 @@ namespace alfi::poly {
 	}
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
-	Container<Number> imp_lagrange(const auto& X, const auto& Y) {
+	Container<Number> imp_lagrange(const Container<Number>& X, const Container<Number>& Y) {
 		if (X.size() != Y.size()) {
 			std::cerr << "Error in function " << __FUNCTION__
 					  << ": Vectors X (of size " << X.size()
@@ -183,7 +183,12 @@ namespace alfi::poly {
 	}
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
-	Container<Number> imp_lagrange_vals(const auto& X, const auto& Y, const auto& xx, Number epsilon = std::numeric_limits<Number>::epsilon()) {
+	Container<Number> imp_lagrange_vals(
+			const Container<Number>& X,
+			const Container<Number>& Y,
+			const Container<Number>& xx,
+			Number epsilon = std::numeric_limits<Number>::epsilon()
+	) {
 		const auto nn = xx.size();
 
 		if (X.size() != Y.size()) {
@@ -241,7 +246,7 @@ namespace alfi::poly {
 	}
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
-	Container<Number> newton(const auto& X, const auto& Y) {
+	Container<Number> newton(const Container<Number>& X, const Container<Number>& Y) {
 		if (X.size() != Y.size()) {
 			std::cerr << "Error in function " << __FUNCTION__
 					  << ": Vectors X (of size " << X.size()
@@ -294,7 +299,7 @@ namespace alfi::poly {
 	}
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
-	Container<Number> newton_vals(const auto& X, const auto& Y, const auto& xx) {
+	Container<Number> newton_vals(const Container<Number>& X, const Container<Number>& Y, const Container<Number>& xx) {
 		const auto nn = xx.size();
 
 		if (X.size() != Y.size()) {

--- a/ALFI/ALFI/spline/cubic.h
+++ b/ALFI/ALFI/spline/cubic.h
@@ -87,7 +87,7 @@ namespace alfi::spline {
 								  typename Types::NotAKnotEnd,
 								  typename Types::SemiNotAKnot>;
 
-		static Container<Number> compute_coeffs(const auto& X, const auto& Y, Type type = typename Types::Default{}) {
+		static Container<Number> compute_coeffs(const Container<Number>& X, const Container<Number>& Y, Type type = typename Types::Default{}) {
 			constexpr auto FUNCTION = __FUNCTION__;
 
 			if (X.size() != Y.size()) {
@@ -131,7 +131,8 @@ namespace alfi::spline {
 					diag[n-1] = dX[n-2] - dX[n-3];
 					upper[n-1] = NAN;
 					right[n-1] = dX[n-2] / (dX[n-2]+dX[n-3]) * right[n-2];
-					const auto C = util::linalg::tridiag_solve(lower, diag, upper, right);
+					const auto C = util::linalg::tridiag_solve<Number,Container>(
+						std::move(lower), std::move(diag), std::move(upper), std::move(right));
 					for (SizeT i = 0, index = 0; i < n - 1; ++i) {
 						coeffs[index++] = (C[i+1] - C[i]) / (3*dX[i]);
 						coeffs[index++] = C[i];
@@ -216,7 +217,8 @@ namespace alfi::spline {
 					diag[n-1] = 2*dX[n-2];
 					upper[n-1] = NAN;
 					right[n-1] = 3 * (c.df_n - dY[n-2]/dX[n-2]);
-					const auto C = util::linalg::tridiag_solve(lower, diag, upper, right);
+					const auto C = util::linalg::tridiag_solve<Number,Container>(
+						std::move(lower), std::move(diag), std::move(upper), std::move(right));
 					for (SizeT i = 0, index = 0; i < n - 1; ++i) {
 						coeffs[index++] = (C[i+1] - C[i]) / (3*dX[i]);
 						coeffs[index++] = C[i];
@@ -241,7 +243,8 @@ namespace alfi::spline {
 					diag[n-1] = 1;
 					upper[n-1] = NAN;
 					right[n-1] = s.d2f_n / 2;
-					const auto C = util::linalg::tridiag_solve(lower, diag, upper, right);
+					const auto C = util::linalg::tridiag_solve<Number,Container>(
+						std::move(lower), std::move(diag), std::move(upper), std::move(right));
 					for (SizeT i = 0, index = 0; i < n - 1; ++i) {
 						coeffs[index++] = (C[i+1] - C[i]) / (3*dX[i]);
 						coeffs[index++] = C[i];
@@ -271,7 +274,8 @@ namespace alfi::spline {
 						diag[n-1] = 1;
 						upper[n-1] = NAN;
 						right[n-1] = dX[n-2] * t.d3f_n / 2;
-						C = util::linalg::tridiag_solve<Number,Container>(lower, diag, upper, right);
+						C = util::linalg::tridiag_solve<Number,Container>(
+							std::move(lower), std::move(diag), std::move(upper), std::move(right));
 					}
 					Container<Number> D(n - 1);
 					for (SizeT i = 1; i < D.size() - 1; ++i) {
@@ -303,7 +307,7 @@ namespace alfi::spline {
 						const auto h12 = X[2]-X[0], h13 = X[3]-X[0];
 						const auto d1 = Y[1]-Y[0], d12 = Y[2]-Y[0], d13 = Y[3]-Y[0];
 
-						const auto abc = util::linalg::lup_solve(
+						const auto abc = util::linalg::lup_solve<Number,Container>(
 							{{std::pow(h1, 3), std::pow(h1, 2), h1},
 								{std::pow(h12, 3), std::pow(h12, 2), h12},
 								{std::pow(h13, 3), std::pow(h13, 2), h13}},
@@ -350,7 +354,7 @@ namespace alfi::spline {
 						const auto h12 = X[n-2]-X[n-4], h13 = X[n-1]-X[n-4];
 						const auto d1 = dY[n-4], d12 = Y[n-2] - Y[n-4], d13 = Y[n-1] - Y[n-4];
 
-						const auto abc = util::linalg::lup_solve(
+						const auto abc = util::linalg::lup_solve<Number,Container>(
 							{{std::pow(h1, 3), std::pow(h1, 2), h1},
 								{std::pow(h12, 3), std::pow(h12, 2), h12},
 								{std::pow(h13, 3), std::pow(h13, 2), h13}},
@@ -396,7 +400,7 @@ namespace alfi::spline {
 		CubicSpline() = default;
 
 		template <typename ContainerXType>
-		CubicSpline(ContainerXType&& X, const auto& Y, Type type = typename Types::Default{}) {
+		CubicSpline(ContainerXType&& X, const Container<Number>& Y, Type type = typename Types::Default{}) {
 			construct(std::forward<ContainerXType>(X), Y, type);
 		}
 
@@ -407,7 +411,7 @@ namespace alfi::spline {
 		CubicSpline& operator=(CubicSpline&& other) noexcept = default;
 
 		template <typename ContainerXType>
-		void construct(ContainerXType&& X, const auto& Y, Type type = typename Types::Default{}) {
+		void construct(ContainerXType&& X, const Container<Number>& Y, Type type = typename Types::Default{}) {
 			if (X.size() != Y.size()) {
 				std::cerr << "Error in function " << __FUNCTION__
 						  << ": Vectors X (of size " << X.size()

--- a/ALFI/ALFI/spline/linear.h
+++ b/ALFI/ALFI/spline/linear.h
@@ -11,7 +11,7 @@ namespace alfi::spline {
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
 	class LinearSpline {
 	public:
-		static Container<Number> compute_coeffs(const auto& X, const auto& Y) {
+		static Container<Number> compute_coeffs(const Container<Number>& X, const Container<Number>& Y) {
 			if (X.size() != Y.size()) {
 				std::cerr << "Error in function " << __FUNCTION__
 						  << ": Vectors X (of size " << X.size()
@@ -47,7 +47,7 @@ namespace alfi::spline {
 		LinearSpline() = default;
 
 		template <typename ContainerXType>
-		LinearSpline(ContainerXType&& X, const auto& Y) {
+		LinearSpline(ContainerXType&& X, const Container<Number>& Y) {
 			construct(std::forward<ContainerXType>(X), Y);
 		}
 
@@ -58,7 +58,7 @@ namespace alfi::spline {
 		LinearSpline& operator=(LinearSpline&& other) noexcept = default;
 
 		template <typename ContainerXType>
-		void construct(ContainerXType&& X, const auto& Y) {
+		void construct(ContainerXType&& X, const Container<Number>& Y) {
 			if (X.size() != Y.size()) {
 				std::cerr << "Error in function " << __FUNCTION__
 						  << ": Vectors X (of size " << X.size()

--- a/ALFI/ALFI/spline/quadratic.h
+++ b/ALFI/ALFI/spline/quadratic.h
@@ -134,7 +134,7 @@ namespace alfi::spline {
 								  typename Types::FixedSecond,
 								  typename Types::NotAKnot>;
 
-		static Container<Number> compute_coeffs(const auto& X, const auto& Y, Type type = typename Types::Default{}) {
+		static Container<Number> compute_coeffs(const Container<Number>& X, const Container<Number>& Y, Type type = typename Types::Default{}) {
 			if (X.size() != Y.size()) {
 				std::cerr << "Error in function " << __FUNCTION__
 						  << ": Vectors X (of size " << X.size()
@@ -265,7 +265,7 @@ namespace alfi::spline {
 		QuadraticSpline() = default;
 
 		template <typename ContainerXType>
-		QuadraticSpline(ContainerXType&& X, const auto& Y, Type type = typename Types::Default{}) {
+		QuadraticSpline(ContainerXType&& X, const Container<Number>& Y, Type type = typename Types::Default{}) {
 			construct(std::forward<ContainerXType>(X), Y, type);
 		}
 
@@ -276,7 +276,7 @@ namespace alfi::spline {
 		QuadraticSpline& operator=(QuadraticSpline&& other) noexcept = default;
 
 		template <typename ContainerXType>
-		void construct(ContainerXType&& X, const auto& Y, Type type = typename Types::Default{}) {
+		void construct(ContainerXType&& X, const Container<Number>& Y, Type type = typename Types::Default{}) {
 			if (X.size() != Y.size()) {
 				std::cerr << "Error in function " << __FUNCTION__
 						  << ": Vectors X (of size " << X.size()

--- a/ALFI/ALFI/util/arrays.h
+++ b/ALFI/ALFI/util/arrays.h
@@ -40,7 +40,7 @@ namespace alfi::util::arrays {
 	}
 
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
-	Container<Number> diff(const auto& container) {
+	Container<Number> diff(const Container<Number>& container) {
 		if (container.empty()) {
 			return {};
 		}

--- a/ALFI/ALFI/util/linalg.h
+++ b/ALFI/ALFI/util/linalg.h
@@ -98,7 +98,12 @@ namespace alfi::util::linalg {
 		@return the solution vector
 	 */
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
-	Container<Number> tridiag_solve_unstable(const auto& lower, auto&& diag, const auto& upper, auto&& right) {
+	Container<Number> tridiag_solve_unstable(
+			const Container<Number>& lower,
+			Container<Number>&& diag,
+			const Container<Number>& upper,
+			Container<Number>&& right
+	) {
 		const auto n = right.size();
 		assert(n == lower.size());
 		assert(n == diag.size());
@@ -140,7 +145,12 @@ namespace alfi::util::linalg {
 		@return the solution vector
 	 */
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
-	Container<Number> tridiag_solve(auto&& lower, auto&& diag, auto&& upper, auto&& right) {
+	Container<Number> tridiag_solve(
+			Container<Number>&& lower,
+			Container<Number>&& diag,
+			Container<Number>&& upper,
+			Container<Number>&& right
+	) {
 		const auto n = right.size();
 		assert(n == lower.size());
 		assert(n == diag.size());

--- a/ALFI/ALFI/util/points.h
+++ b/ALFI/ALFI/util/points.h
@@ -16,7 +16,7 @@ namespace alfi::points {
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
 	Container<Number> lin_mapped(const Container<Number>& points, Number a, Number b, Number c, Number d) {
 		auto mapped_points = points;
-		lin_map(mapped_points, a, b, c, d);
+		lin_map<Number,Container>(mapped_points, a, b, c, d);
 		return mapped_points;
 	}
 
@@ -29,7 +29,7 @@ namespace alfi::points {
 			std::fill(points.begin(), points.end(), (a + b) / 2);
 			return;
 		}
-		lin_map(points, points.front(), points.back(), a, b);
+		lin_map<Number,Container>(points, points.front(), points.back(), a, b);
 		points.front() = a;
 		points.back() = b;
 	}
@@ -37,7 +37,7 @@ namespace alfi::points {
 	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
 	Container<Number> stretched(const Container<Number>& points, Number a, Number b) {
 		auto stretched_points = points;
-		stretch(stretched_points, a, b);
+		stretch<Number,Container>(stretched_points, a, b);
 		return stretched_points;
 	}
 }

--- a/ALFI/ALFI/util/points.h
+++ b/ALFI/ALFI/util/points.h
@@ -3,8 +3,8 @@
 #include "../config.h"
 
 namespace alfi::points {
-	template <typename Number = DefaultNumber>
-	void lin_map(auto& points, Number a, Number b, Number c, Number d) {
+	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
+	void lin_map(Container<Number>& points, Number a, Number b, Number c, Number d) {
 		const auto mid1 = (a + b) / 2;
 		const auto mid2 = (c + d) / 2;
 		const auto scale = (d - c) / (b - a);
@@ -13,15 +13,15 @@ namespace alfi::points {
 		}
 	}
 
-	template <typename Number = DefaultNumber>
-	auto lin_mapped(const auto& points, Number a, Number b, Number c, Number d) {
+	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
+	Container<Number> lin_mapped(const Container<Number>& points, Number a, Number b, Number c, Number d) {
 		auto mapped_points = points;
 		lin_map(mapped_points, a, b, c, d);
 		return mapped_points;
 	}
 
-	template <typename Number = DefaultNumber>
-	void stretch(auto& points, Number a, Number b) {
+	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
+	void stretch(Container<Number>& points, Number a, Number b) {
 		if (points.empty()) {
 			return;
 		}
@@ -34,8 +34,8 @@ namespace alfi::points {
 		points.back() = b;
 	}
 
-	template <typename Number = DefaultNumber>
-	auto stretched(const auto& points, Number a, Number b) {
+	template <typename Number = DefaultNumber, template <typename> class Container = DefaultContainer>
+	Container<Number> stretched(const Container<Number>& points, Number a, Number b) {
 		auto stretched_points = points;
 		stretch(stretched_points, a, b);
 		return stretched_points;

--- a/ALFI/ALFI/util/spline.h
+++ b/ALFI/ALFI/util/spline.h
@@ -65,7 +65,7 @@ namespace alfi::util::spline {
 			const auto h12 = X[2]-X[0], h13 = X[3]-X[0];
 			const auto d1 = Y[1]-Y[0], d12 = Y[2]-Y[0], d13 = Y[3]-Y[0];
 
-			const auto abc = util::linalg::lup_solve(
+			const auto abc = linalg::lup_solve<Number,Container>(
 				{{std::pow(h1, 3), std::pow(h1, 2), h1},
 					{std::pow(h12, 3), std::pow(h12, 2), h12},
 					{std::pow(h13, 3), std::pow(h13, 2), h13}},

--- a/examples/dist_QCustomPlot/CMakeLists.txt
+++ b/examples/dist_QCustomPlot/CMakeLists.txt
@@ -1,4 +1,4 @@
-set(CMAKE_CXX_STANDARD 23)
+set(CMAKE_CXX_STANDARD 17)
 
 find_package(Qt5 REQUIRED COMPONENTS Widgets)
 find_package(QCustomPlot REQUIRED)

--- a/examples/dist_gnuplot/CMakeLists.txt
+++ b/examples/dist_gnuplot/CMakeLists.txt
@@ -1,4 +1,4 @@
-set(CMAKE_CXX_STANDARD 23)
+set(CMAKE_CXX_STANDARD 17)
 
 find_package(Boost REQUIRED COMPONENTS iostreams)
 find_package(Gnuplot REQUIRED)

--- a/examples/interpolation/CMakeLists.txt
+++ b/examples/interpolation/CMakeLists.txt
@@ -1,4 +1,4 @@
-set(CMAKE_CXX_STANDARD 23)
+set(CMAKE_CXX_STANDARD 20)
 
 find_package(Qt5 REQUIRED COMPONENTS Widgets)
 find_package(QCustomPlot REQUIRED)

--- a/examples/interpolation/interpolation.cpp
+++ b/examples/interpolation/interpolation.cpp
@@ -30,11 +30,11 @@ class PlotWindow final : public QWidget {
 public:
 	static const inline std::vector<std::pair<QString,std::function<std::vector<double>(std::vector<double>,std::vector<double>,std::vector<double>)>>>
 	poly_types = {
-		{"Lagrange", [](const auto& X, const auto& Y, const auto& xx) { return alfi::poly::val(alfi::poly::lagrange(X, Y), xx); }},
+		{"Lagrange", [](const auto& X, const auto& Y, const auto& xx) { return alfi::poly::val<double,std::vector>(alfi::poly::lagrange(X, Y), xx); }},
 		{"Lagrange values", [](const auto& X, const auto& Y, const auto& xx) { return alfi::poly::lagrange_vals(X, Y, xx); }},
-		{"Improved Lagrange", [](const auto& X, const auto& Y, const auto& xx) { return alfi::poly::val(alfi::poly::imp_lagrange(X, Y), xx); }},
+		{"Improved Lagrange", [](const auto& X, const auto& Y, const auto& xx) { return alfi::poly::val<double,std::vector>(alfi::poly::imp_lagrange(X, Y), xx); }},
 		{"Improved Lagrange values", [](const auto& X, const auto& Y, const auto& xx) { return alfi::poly::imp_lagrange_vals(X, Y, xx); }},
-		{"Newton", [](const auto& X, const auto& Y, const auto& xx) { return alfi::poly::val(alfi::poly::newton(X, Y), xx); }},
+		{"Newton", [](const auto& X, const auto& Y, const auto& xx) { return alfi::poly::val<double,std::vector>(alfi::poly::newton(X, Y), xx); }},
 		{"Newton values", [](const auto& X, const auto& Y, const auto& xx) { return alfi::poly::newton_vals(X, Y, xx); }},
 	};
 	static const inline size_t poly_default_type_index = 0;
@@ -332,19 +332,19 @@ private:
 			if (barycentric_dist_type == 0) {
 				barycentric_dist_type = dist_type;
 			}
-			add_graph("Barycentric", xx, alfi::misc::barycentric(X, Y, xx, static_cast<alfi::dist::Type>(barycentric_dist_type)));
+			add_graph("Barycentric", xx, alfi::misc::barycentric<double,std::vector>(X, Y, xx, static_cast<alfi::dist::Type>(barycentric_dist_type)));
 		}
 		if (_step_spline_checkbox->isChecked()) {
 			add_graph("Step Spline", xx, alfi::spline::StepSpline(X, Y, step_spline_types[_step_spline_combo->currentIndex()].second)(xx));
 		}
 		if (_linear_spline_checkbox->isChecked()) {
-			add_graph("Linear Spline", xx, alfi::spline::LinearSpline(X, Y)(xx));
+			add_graph("Linear Spline", xx, alfi::spline::LinearSpline<>(X, Y)(xx));
 		}
 		if (_quadratic_spline_checkbox->isChecked()) {
-			add_graph("Quadratic Spline", xx, alfi::spline::QuadraticSpline(X, Y, quadratic_spline_types[_quadratic_spline_combo->currentIndex()].second)(xx));
+			add_graph("Quadratic Spline", xx, alfi::spline::QuadraticSpline<>(X, Y, quadratic_spline_types[_quadratic_spline_combo->currentIndex()].second)(xx));
 		}
 		if (_cubic_spline_checkbox->isChecked()) {
-			add_graph("Cubic Spline", xx, alfi::spline::CubicSpline(X, Y, cubic_spline_types[_cubic_spline_combo->currentIndex()].second)(xx));
+			add_graph("Cubic Spline", xx, alfi::spline::CubicSpline<>(X, Y, cubic_spline_types[_cubic_spline_combo->currentIndex()].second)(xx));
 		}
 
 		if (_default_axis_ranges) {

